opentracing-api/src/main/java/io/opentracing/propagation/Binary.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2016-2017 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.opentracing.propagation;
+
+import java.io.IOException;
+
+/**
+ * Binary is an interface defining the required operations for a binary carrier for
+ * Tracer.inject() and Tracer.extract(). Binary can be defined either as inbound (extraction)
+ * or outbound (injection).
+ *
+ * When Binary is defined as inbound, read() must be used to read data,
+ * and it is an error to call read() or getOutboundPayload().
+ *
+ * When Binary is defined as outbound, write() must be used to write data,
+ * getOutboundPayload() to retrieve all the written data, and it is an error to call read().
+ *
+ * @see Format.Builtin#BINARY
+ * @see io.opentracing.Tracer#inject(SpanContext, Format, Object)
+ * @see io.opentracing.Tracer#extract(Format, Object)
+ */
+public interface Binary {
+    /**
+     * @return whether this instance is inbound and supports reading.
+     */
+    boolean isInbound();
+
+    /**
+     * @return whether this instance is outbound and supports writing.
+     */
+    boolean isOutbound();
+
+    /**
+     * Writes b.length bytes from the specified byte array. The internal
+     * buffer should grow as data is written to it.
+     *
+     * @param b the data.
+     */
+    void write(byte[] b) throws IOException;
```

Comment:
I don't have a strong opinion about the exact read/write API, but would it make more sense to model these functions after nio [ReadableByteChannel](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/ReadableByteChannel.html) and [WritableByteChannel](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/WritableByteChannel.html)? They each have just a single method, but it's based on ByteBuffer, which I'm guessing makes them equivalent to several overloaded methods of Input/OutputStream, and certainly more flexible than `write(byte[] b)` (which requires an allocation if you're writing in batches of different sizes). Maybe it's not an issue for propagation codecs specifically.

Reply:
Agreed. ByteBuffer provides a greater level of flexibility while still being able to wrap (without copy) a byte[] with ByteBuffer.wrap